failing installation of GCCcore/10.3.0 on top of EESSI 2023.04 compat layer

The installation of `GCCcore/10.3.0` is failing with EasyBuild v4.7.1 on top of the EESSI pilot 2023.04 compat layer, with:

```
/tmp/boegel/easybuild/build/GCCcore/10.3.0/system-system/gcc-10.3.0/stage1_obj/./gcc/xgcc -B/tmp/boegel/easybuild/build/GCCcore/10.3.0/system-system/gcc-10.3.0/stage1_obj/./gcc/ -B/tmp/boegel/easybuild/build/GCCcore/10.3.0/system-system/GCC_stage1_eb/x86_64-pc-linux-gnu/bin/ -B/tmp/boegel/easybuild/build/GCCcore/10.3.0/system-system/GCC_stage1_eb/x86_64-pc-linux-gnu/lib/ -isystem /tmp/boegel/easybuild/build/GCCcore/10.3.0/system-system/GCC_stage1_eb/x86_64-pc-linux-gnu/include -isystem /tmp/boegel/easybuild/build/GCCcore/10.3.0/system-system/GCC_stage1_eb/x86_64-pc-linux-gnu/sys-include    -O2  -g -O2 -DIN_GCC    -W -Wall -Wno-narrowing -Wwrite-strings -Wcast-qual -Wstrict-prototypes -Wmissing-prototypes -Wold-style-definition  -isystem ./include  -fpic -mlong-double-80 -DUSE_ELF_SYMVER  -g -DIN_LIBGCC2 -fbuilding-libgcc -fno-stack-protector  -shared -nodefaultlibs -Wl,--soname=libgcc_s.so.1 -Wl,--version-script=libgcc.map -o ./libgcc_s.so.1.tmp -g -O2 -B./ _muldi3_s.o _negdi2_s.o _lshrdi3_s.o _ashldi3_s.o _ashrdi3_s.o _cmpdi2_s.o _ucmpdi2_s.o _clear_cache_s.o _trampoline_s.o __main_s.o _absvsi2_s.o _absvdi2_s.o _addvsi3_s.o _addvdi3_s.o _subvsi3_s.o _subvdi3_s.o _mulvsi3_s.o _mulvdi3_s.o _negvsi2_s.o _negvdi2_s.o _ctors_s.o _ffssi2_s.o _ffsdi2_s.o _clz_s.o _clzsi2_s.o _clzdi2_s.o _ctzsi2_s.o _ctzdi2_s.o _popcount_tab_s.o _popcountsi2_s.o _popcountdi2_s.o _paritysi2_s.o _paritydi2_s.o _powisf2_s.o _powidf2_s.o _powixf2_s.o _mulhc3_s.o _mulsc3_s.o _muldc3_s.o _mulxc3_s.o _divhc3_s.o _divsc3_s.o _divdc3_s.o _divxc3_s.o _bswapsi2_s.o _bswapdi2_s.o _clrsbsi2_s.o _clrsbdi2_s.o _fixunssfsi_s.o _fixunsdfsi_s.o _fixunsxfsi_s.o _fixsfdi_s.o _fixdfdi_s.o _fixxfdi_s.o _fixunssfdi_s.o _fixunsdfdi_s.o _fixunsxfdi_s.o _floatdisf_s.o _floatdidf_s.o _floatdixf_s.o _floatundisf_s.o _floatundidf_s.o _floatundixf_s.o _divdi3_s.o _moddi3_s.o _divmoddi4_s.o _udivdi3_s.o _umoddi3_s.o _udivmoddi4_s.o _udiv_w_sdiv_s.o cpuinfo_s.o sfp-exceptions_s.o addtf3_s.o divtf3_s.o multf3_s.o negtf2_s.o subtf3_s.o unordtf2_s.o fixtfsi_s.o fixunstfsi_s.o floatsitf_s.o floatunsitf_s.o fixtfdi_s.o fixunstfdi_s.o floatditf_s.o floatunditf_s.o fixtfti_s.o fixunstfti_s.o floattitf_s.o floatuntitf_s.o extendsftf2_s.o extenddftf2_s.o extendxftf2_s.o trunctfsf2_s.o trunctfdf2_s.o trunctfxf2_s.o getf2_s.o letf2_s.o eqtf2_s.o _divtc3_s.o _multc3_s.o _powitf2_s.o enable-execute-stack_s.o unwind-dw2_s.o unwind-dw2-fde-dip_s.o unwind-sjlj_s.o unwind-c_s.o emutls_s.o libgcc.a -lc && rm -f ./libgcc_s.so && if [ -f ./libgcc_s.so.1 ]; then mv -f ./libgcc_s.so.1 ./libgcc_s.so.1.backup; else true; fi && mv ./libgcc_s.so.1.tmp ./libgcc_s.so.1 && (echo "/* GNU ld script"; echo "   Use the shared library, but some functions are only in"; echo "   the static library.  */"; echo "GROUP ( libgcc_s.so.1 -lgcc )" ) > ./libgcc_s.so
/cvmfs/pilot.eessi-hpc.org/versions/2023.04/compat/linux/x86_64/usr/bin/ld.gold: error: cannot open /lib64/libc.so.6: No such file or directory
/cvmfs/pilot.eessi-hpc.org/versions/2023.04/compat/linux/x86_64/usr/bin/ld.gold: error: cannot open /usr/lib64/libc_nonshared.a: No such file or directory
collect2: error: ld returned 1 exit status
make[2]: *** [Makefile:994: libgcc_s.so] Error 1
make[2]: Leaving directory '/tmp/boegel/easybuild/build/GCCcore/10.3.0/system-system/gcc-10.3.0/stage1_obj/x86_64-pc-linux-gnu/libgcc'
make[1]: *** [Makefile:14332: all-target-libgcc] Error 2
make[1]: Leaving directory '/tmp/boegel/easybuild/build/GCCcore/10.3.0/system-system/gcc-10.3.0/stage1_obj'
make: *** [Makefile:967: all] Error 2
```

see also https://github.com/NorESSI/software-layer/pull/96 (cc @trz42)

--- a/eb_hooks.py
+++ b/eb_hooks.py
@@ -3,8 +3,11 @@
 import os
 import re
 
+from easybuild.easyblocks.generic.configuremake import obtain_config_guess
 from easybuild.tools.build_log import EasyBuildError, print_msg
 from easybuild.tools.config import build_option, update_build_option
+from easybuild.tools.filetools import copy_file, which
+from easybuild.tools.run import run_cmd
 from easybuild.tools.systemtools import AARCH64, POWER, X86_64, get_cpu_architecture, get_cpu_features
 from easybuild.tools.toolchain.compiler import OPTARCH_GENERIC
 
@@ -89,6 +92,25 @@ def pre_prepare_hook(self, *args, **kwargs):
                   mpi_family, rpath_override_dirs)
 
 
+def gcc_postprepare(self, *args, **kwargs):
+    """
+    Post-configure hook for GCCcore:
+    - copy RPATH wrapper script for linker commands to also have a wrapper in place with system type prefix like 'x86_64-pc-linux-gnu'
+    """
+    if self.name == 'GCCcore':
+        config_guess = obtain_config_guess()
+        system_type, _ = run_cmd(config_guess, log_all=True)
+        cmd_prefix = '%s-' % system_type.strip()
+        for cmd in ('ld', 'ld.gold', 'ld.bfd'):
+            wrapper = which(cmd)
+            self.log.info("Path to %s wrapper: %s" % (cmd, wrapper))
+            wrapper_dir = os.path.dirname(wrapper)
+            prefix_wrapper = os.path.join(wrapper_dir, cmd_prefix + cmd)
+            copy_file(wrapper, prefix_wrapper)
+            self.log.info("Path to %s wrapper with '%s' prefix: %s" % (cmd, cmd_prefix, which(prefix_wrapper)))
+    else:
+        raise EasyBuildError("GCCcore-specific hook triggered for non-GCCcore easyconfig?!")
+
 def post_prepare_hook(self, *args, **kwargs):
     """Main post-prepare hook: trigger custom functions."""
 
@@ -97,6 +119,9 @@ def post_prepare_hook(self, *args, **kwargs):
         update_build_option('rpath_override_dirs', getattr(self, EESSI_RPATH_OVERRIDE_ATTR))
         print_msg("Resetting rpath_override_dirs to original value: %s", getattr(self, EESSI_RPATH_OVERRIDE_ATTR))
         delattr(self, EESSI_RPATH_OVERRIDE_ATTR)
+
+    if self.name in POST_PREPARE_HOOKS:
+        POST_PREPARE_HOOKS[self.name](self, *args, **kwargs)
 
 
 def cgal_toolchainopts_precise(ec, eprefix):
@@ -185,6 +210,10 @@ PARSE_HOOKS = {
     'CGAL': cgal_toolchainopts_precise,
     'fontconfig': fontconfig_add_fonts,
     'UCX': ucx_eprefix,
+}
+
+POST_PREPARE_HOOKS = {
+    'GCCcore': gcc_postprepare,
 }
 
 PRE_CONFIGURE_HOOKS = {

--- a/eb_hooks.py
+++ b/eb_hooks.py
@@ -6,7 +6,7 @@ import re
 from easybuild.easyblocks.generic.configuremake import obtain_config_guess
 from easybuild.tools.build_log import EasyBuildError, print_msg
 from easybuild.tools.config import build_option, update_build_option
-from easybuild.tools.filetools import copy_file, which
+from easybuild.tools.filetools import apply_regex_substitutions, copy_file, which
 from easybuild.tools.run import run_cmd
 from easybuild.tools.systemtools import AARCH64, POWER, X86_64, get_cpu_architecture, get_cpu_features
 from easybuild.tools.toolchain.compiler import OPTARCH_GENERIC
@@ -108,6 +108,17 @@ def gcc_postprepare(self, *args, **kwargs):
             prefix_wrapper = os.path.join(wrapper_dir, cmd_prefix + cmd)
             copy_file(wrapper, prefix_wrapper)
             self.log.info("Path to %s wrapper with '%s' prefix: %s" % (cmd, cmd_prefix, which(prefix_wrapper)))
+
+            # we need to tweak the copied wrapper script, so that:
+            regex_subs = [
+                # - CMD in the script is set to the command name without prefix, because EasyBuild's rpath_args.py
+                #   script that is used by the wrapper script only checks for 'ld', 'ld.gold', etc.
+                #   when checking whether or not to use -Wl
+                ('^CMD=.*', 'CMD=%s' % cmd),
+                # - the path to the correct actual binary is logged and called
+                ('/%s ' % cmd, '/%s ' % (cmd_prefix + cmd)),
+            ]
+            apply_regex_substitutions(prefix_wrapper, regex_subs)
     else:
         raise EasyBuildError("GCCcore-specific hook triggered for non-GCCcore easyconfig?!")
 

